feat(apple): Replay Redact (#11418)
getsentry/sentry-docs: docs/platforms/apple/guides/ios/session-replay/customredact.mdx

---

> title: Using Custom Redact for Session Replay

Reviewer: Should we name it `Privacy` similar to the JS sdk? https://docs.sentry.io/platforms/javascript/session-replay/privacy

Author: I dont know, should we?

Reviewer: I would also add here the `resume/pause` functionality, and I think in this case it'd make sense to rename it as it's not only custom redaction/masking anymore. Maybe `Privacy Controls` or something like that would do?

Author: Good point. I will improve in another doc. This is waiting for too long

sidebar_order: 5501
notSupported:

> description: "Learn how to redact parts of your app's data in Session Replay."

Reviewer: ```suggestion
description: "Learn how to mask parts of your app's data in Session Replay."
```
i guess we have to adapt the naming here after changing it in the sdk as well :)

---

<Note>

Using custom redaction for Session Replay can expose sensitive data. Make sure to double-check every part of your app's data that you choose to redact or not.

</Note>

> By default, our Session Replay SDK masks all text content, images, and user input, giving you heightened confidence that no sensitive data will leave the device. However, you can choose which parts of your app's data to redact or not by using some different options.
> 

Reviewer 2: ```suggestion
By default, our Session Replay SDK masks all text content, images, and user input. This helps ensure that no sensitive data will be exposed. You can also manually choose which parts of your app's data to redact by using the different options listed below.

```

> ## Redact by View class

Reviewer 2: ```suggestion
## Redact by View Class
```

> You can choose which type of view you want to redact or ignore by using the `redactViewClasses` or `ignoreViewClasses` options.

Reviewer 3: `h`: After reading this and the whole document, I'm still unsure of the difference between redact and ignore. Can we add a short explanation, please?

Author: The document was updated. Please take another look

> Let's say you have a custom view that you want to redact and a UILabel subclass (which normally would be redacted) that you want to ignore. You can set the options like this:
> 

Reviewer 2: ```suggestion
Let's say you have a custom view that you want to redact and a `UILabel` subclass (which normally would be redacted) that you want to ignore. You can set the options like this:

```

```swift
  options.experimental.sessionReplay.redactViewClasses = [MyCustomView.self]
  options.experimental.sessionReplay.ignoreViewClasses = [MyCustomLabel.self]

> ```

Reviewer 3: `m`: What drove you to skip adding code samples for ObjC?

Author: Im just following the existing Session replay docs that already dont include objc sample

> ## Redact View by instance

Reviewer 2: ```suggestion
## Redact by View Instance
```

> You can also choose to redact or ignore a specific view instance by using the [replay API](/api/replays/) or view extensions like this:

Reviewer: ```suggestion
You can also choose to redact or ignore a specific view instance by using the public `SentrySDK.replay` API or view extensions like this:
```
i think this link to the api page is not what we meant

> You can also choose to redact or ignore a specific view instance by using the replay API or view extensions like this:

Reviewer 2: ```suggestion
You can also choose to redact or ignore a specific view instance by using the [replay API](/api/replays/) or view extensions like this:
```

```swift
  SentrySDK.replay.redactView(view: view)
  SentrySDK.replay.ignoreView(view: label)
```
or

```swift
  view.sentryReplayRedact()
  label.sentryReplayIgnore()
```

## SwiftUI

> Because of the way SwiftUI is transformed into UIKit, it will be often over-redacted. Modifier like `background` use the same element as an `Image`.

Reviewer 4: (nitpicks on grammar)
- "...it will **be often** over-redacted" -> "it will **often be** over-redacted"
- "Modifier ..." -> "Modifier**s** ..."    or   "A modifier ... use**s** the same ..."  (not sure which one is meant here)

Author: ```suggestion
Because of the way SwiftUI is transformed into UIKit, it will often be over-redacted. A modifier like `background` uses the same element as an `Image`.
```

> In order to be able to control SwiftUI redact process, first you need to disable the default redaction options:

Reviewer 4: (nitpick on grammar)
- "In order to **be able to control** SwiftUI redact process..." -> "In order to control **the** SwiftUI redact process..."

Reviewer 4: Also (speaking as a total novice) it is not 100% clear if I need to disable the default options to be able to ignore view instances as described above (my intuition would say no, so even with default redaction I can manually un-redact views)

Author: ```suggestion
In order to control the SwiftUI redact process, you first need to disable the default redaction options:
```

```swift
    options.experimental.sessionReplay.redactAllText = false
    options.experimental.sessionReplay.redactAllImages = false
```

> Then you can manually choose which `View` you want to redact with the `replayRedact` modifier

Reviewer 2: ```suggestion
Then you can manually choose which `View` you want to redact with the `replayRedact` modifier:
```

```swift    
    var body: some View {
        Text("Hello, World!")

>             .sentryReplayRedact()

Reviewer 3: `m`: Can't you also wrap this with something like as you can for `SentryTracedView`?
```
SentryReplayRedact {
    Text("Hello world")

}
```

If not, maybe we should add it for consistency.

Author: No, modifiers don’t work like that. Technically, it would be possible to create a view that wraps another view, but it would be a hack. My understanding is that the modifier approach is the preferred way in SwiftUI.

    }
```

> <Note>
> Dissabling the default redaction options will expose all text and images in the session replay. 
> Make sure to redact them manually.
> </Note>

Reviewer 2: ```suggestion
<Note>
Disabling the default redaction options will expose all text and images in the session replay. 
Make sure to redact them manually.
</Note>
```

---
title: Using Custom Masking for Session Replay
sidebar_order: 5501
notSupported:
description: "Learn how to mask parts of your app's data in Session Replay."
---

<Alert>

> 

Author: ```suggestion
Before publising an App with Session Replay enabled, make sure to test it thoroughly to ensure that no sensitive data is exposed.

```

> If you choose to use custom masking in your Session Replays, you may accidentally expose sensitive customer data. Be sure to double-check what you choose to expose.

Reviewer 5: That's because for that they must disable the default masking, right?
Worth calling that out?

@lizokm might have ideas?

</Alert>

By default, our Session Replay SDK masks all text content, images, and user input. This helps ensure that no sensitive data will be exposed. You can also manually choose which parts of your app's data you want to mask by using the different options listed below.

## Mask by View Class

> You can choose which type of view you want to mask or ignore by using the `maskedViewClasses` or `unmaskedViewClasses` options.

Reviewer 3: `m`: Maybe putting it like this makes it easier to understand.
```suggestion
You can choose which type of view you want to mask or unmask by using the `maskedViewClasses` or `unmaskedViewClasses` options.
```

> Let's say you have a custom view that you want to mask and a `UILabel` subclass (which normally would be masked) that you doesn't want to mask. You can set the options like this:

Reviewer 3: ```suggestion
Let's say you have a custom view that you want to mask and a `UILabel` subclass (which normally would be masked) that you don't want to mask. You can set the options like this:
```

```swift
  options.experimental.sessionReplay.maskedViewClasses = [MyCustomView.self]
  options.experimental.sessionReplay.unmaskedViewClasses = [MyCustomLabel.self]
```

## Mask by View Instance

> You can also choose to mask or unmask a specific view instance by using the [replay API](/api/replays/) or view extensions like this:

Author: ```suggestion
You can also choose to mask or unmask a specific view instance by using the replay API (SentrySDK.replay) or view extensions like this:
```

```swift
  SentrySDK.replay.maskView(view: view)
  SentrySDK.replay.unmaskView(view: label)
```
or

```swift
  view.sentryReplayMask()
  label.sentryReplayUnmask()
```

## SwiftUI

Because of the way SwiftUI is transformed into UIKit, it will often be over-masked. A modifier like `background` uses the same element as an `Image`.
In order to control the SwiftUI masking process, you need to use the `sentryReplayUnmask` and/or `sentryReplayMask` modifiers. 

In this example we want to show the message, but not the user name.
```swift
  @Binding var user: String

  var body: some View {
    VStack {
      Text("Hello")
        .sentryReplayUnmask()
      Text("\(user)")
    }
  }
```

In this example, we need to unmask the VStack because its background element will be masked by default.
To hide the username, we need to mask it.

```swift
  @Binding var user: String

  var body: some View {
    VStack {
      Text("Hello")
      Text("\(user)")
        .sentrtReplayMask()
    }
    .background(.blue)
    .sentryReplayUnmask() 
  }
```

<Alert>

> Before publising an App with Session Replay enabled, make sure to test it thoroughly to ensure that no sensitive data is exposed.

Reviewer 3: `m`: I think this should be included in the alert at the top of the file.

> <Alert>
> 
> Before publising an App with Session Replay enabled, make sure to test it thoroughly to ensure that no sensitive data is exposed.
> 
> </Alert>

Author: ```suggestion
```
